lib/rules/html-self-closing-style.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * @author Toru Nagashima
+ * @copyright 2016 Toru Nagashima. All rights reserved.
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const utils = require('../utils')
+
+// ------------------------------------------------------------------------------
+// Helpers
+// ------------------------------------------------------------------------------
+
+/**
+ * Kind strings.
+ * This strings wil be displayed in error messages.
+ */
+const KIND = Object.freeze({
+  NORMAL: 'HTML elements',
+  VOID: 'HTML void elements',
+  COMPONENT: 'Vue.js custom components',
+  SVG: 'SVG elements',
+  MATH: 'MathML elements'
+})
+
+/**
+ * Normalize the given options.
+ * @param {Object|undefined} options The raw options object.
+ * @returns {Object} Normalized options.
+ */
+function parseOptions (options) {
+  return {
+    [KIND.NORMAL]: (options && options.html && options.html.normal) || 'never',
+    [KIND.VOID]: (options && options.html && options.html.void) || 'never',
+    [KIND.COMPONENT]: (options && options.html && options.html.component) || 'always',
+    [KIND.SVG]: (options && options.svg) || 'always',
+    [KIND.MATH]: (options && options.math) || 'always'
+  }
+}
+
+/**
+ * Get the kind of the given element.
+ * @param {VElement} node The element node to get.
+ * @returns {string} The kind of the element.
+ */
+function getKind (node) {
+  if (utils.isCustomComponent(node)) {
+    return KIND.COMPONENT
+  }
+  if (utils.isHtmlElementNode(node)) {
+    if (utils.isHtmlVoidElementName(node.name)) {
+      return KIND.VOID
+    }
+    return KIND.NORMAL
+  }
+  if (utils.isSvgElementNode(node)) {
+    return KIND.SVG
+  }
+  if (utils.isMathMLElementNode(node)) {
+    return KIND.MATH
+  }
+  return 'unknown elements'
+}
+
+/**
+ * Check whether the given element is empty or not.
+ * This ignores whitespaces, doesn't ignore comments.
+ * @param {VElement} node The element node to check.
+ * @returns {boolean} `true` if the element is empty.
+ */
+function isEmpty (node, sourceCode) {
+  const start = node.startTag.range[1]
+  const end = (node.endTag != null) ? node.endTag.range[0] : node.range[1]
+
+  return sourceCode.text.slice(start, end).trim() === ''
+}
+
+/**
+ * Creates AST event handlers for html-self-closing-style.
+ *
+ * @param {RuleContext} context - The rule context.
+ * @returns {object} AST event handlers.
+ */
+function create (context) {
+  const sourceCode = context.getSourceCode()
+  const options = parseOptions(context.options[0])
+
+  utils.registerTemplateBodyVisitor(context, {
+    'VElement' (node) {
+      const kind = getKind(node)
```

Comment:
I'd keep `kind` simple as: `normal`, `void`, `component` and so on, and only reference the longer (more descriptive) name in `message` using e.g. `getKindName(kind) {}` function.

Btw. I think we should rename `kind` and call it for example `htmlElementType` or `nodeType`, what do you think? `kind` itself doesn't mean much, kind of what?